syntax = "proto2";

package tests;

message MessageOfTypes {
    extensions 100 to max;

    required double dubl = 1;
    required float flot = 2;
    required int32 i32 = 3;
    required int64 i64 = 4;
    required uint32 ui32 = 5;
    required uint64 ui64 = 6;
    required sint32 si32 = 7;
    required sint64 si64 = 8;
    required fixed32 f32 = 9;
    required fixed64 f64 = 17;  // oops, OOO
    required sfixed32 sf32 = 10;
    required sfixed64 sf64 = 11;
    required bool bol = 12;
    required string strng = 13;
    required bytes byts = 14;
    required NestedType nested = 15;
    required Enum enm = 16;
    repeated int32 range = 18;
    repeated NestedType nestedRepeated = 19;
    repeated Enum enmRepeated = 20;

    message NestedType { required string req = 1; }

    enum Enum {
        A = 0;
        B = 1;
        C = 2;
    }
}

extend MessageOfTypes {
    optional double extDouble = 100;
    optional string extString = 101;
}

message NestedExtension {
    extend MessageOfTypes {
        optional int32 extInt = 102;
        optional MessageOfTypes.NestedType extNested = 103;
    }
}